cosmic-panel-bin/src/space/panel_space.rs

```diff
@@ -92,10 +95,61 @@ pub enum AppletMsg {
     Cleanup(String),
 }
 
-render_elements! {
-    pub(crate) MyRenderElements<R> where R: ImportMem + ImportAll;
-    Memory=MemoryRenderBufferRenderElement<R>,
-    WaylandSurface=WaylandSurfaceRenderElement<R>
+pub(crate) enum PanelRenderElement {
+    Wayland(WaylandSurfaceRenderElement<GlesRenderer>),
+    RoundedRectangle(RoundedRectangleShaderElement),
+}
+
+impl Element for PanelRenderElement {
+    fn id(&self) -> &smithay::backend::renderer::element::Id {
+        match self {
+            Self::Wayland(e) => e.id(),
+            Self::RoundedRectangle(e) => e.id(),
+        }
+    }
+
+    fn current_commit(&self) -> smithay::backend::renderer::utils::CommitCounter {
+        match self {
+            Self::Wayland(e) => e.current_commit(),
+            Self::RoundedRectangle(e) => e.current_commit(),
+        }
+    }
+
+    fn src(&self) -> Rectangle<f64, Buffer> {
+        match self {
+            Self::Wayland(e) => e.src(),
+            Self::RoundedRectangle(e) => e.src(),
+        }
+    }
+
+    fn geometry(&self, scale: smithay::utils::Scale<f64>) -> Rectangle<i32, Physical> {
+        match self {
+            Self::Wayland(e) => e.geometry(scale),
+            Self::RoundedRectangle(e) => e.geometry(scale),
+        }
+    }
+}
+
+impl RenderElement<GlesRenderer> for PanelRenderElement {
+    fn draw(
+        &self,
+        frame: &mut GlesFrame<'_>,
+        src: Rectangle<f64, Buffer>,
+        dst: Rectangle<i32, Physical>,
+        damage: &[Rectangle<i32, Physical>],
+    ) -> Result<(), GlesError> {
+        match self {
+            Self::Wayland(e) => e.draw(frame, src, dst, damage),
+            Self::RoundedRectangle(e) => e.draw(frame, src, dst, damage),
+        }
+    }
+
+    fn underlying_storage(&self, renderer: &mut GlesRenderer) -> Option<UnderlyingStorage> {
+        match self {
+            PanelRenderElement::Wayland(e) => e.underlying_storage(renderer),
+            PanelRenderElement::RoundedRectangle(e) => e.underlying_storage(renderer),
+        }
+    }
```

Comment:
Given you are using a fixed Renderer type, you can probably use smithays `render_elements`-macro to generate all of this.

Reply:
I tried to do so but I couldn't figure out how to resolve the error when adding the RoundedRectangle variant. I'm not sure exactly why but it believed that it didn't meet the trait bounds.